httpclient/api_client.go

```diff
@@ -196,7 +199,12 @@ func (c *ApiClient) attempt(
 		if err != nil {
 			return c.failRequest(ctx, "failed in rate limiter", err)
 		}
-		request, err := http.NewRequestWithContext(ctx, method, requestURL, requestBody.Reader)
+
+		// This custom context enables us to extend the request timeout
+		// while the request or response body is being read.
+		// It exists because the net/http package supports only a fixed timeout.
+		ctx, ticker := newTimeoutContext(ctx, c.config.HTTPTimeout)
+		request, err := http.NewRequestWithContext(ctx, method, requestURL, tickingReader(ticker, requestBody.Reader))
```

Comment:
Ok, dug through the guts of `net/http` and I believe the client uses a 32kb buffer when reading from the request body. So this would mean that if less than 32kb is written out over `HTTPTimeout` duration, the request could be broken. Given the default 30s timeout, the user needs to sustain more than 1066b/s. Sounds good to me.